From 7419190156c809461b884d41f565ee12d29a893e Mon Sep 17 00:00:00 2001
From: BoyOrigin <ahmadyasinfikri@gmail.com>
Date: Tue, 19 Sep 2023 03:27:46 +0700
Subject: [PATCH] Ability to setting the cursor position


diff --git a/src/wl_window.c b/src/wl_window.c
index d9f2fdcc..602981b9 100644
--- a/src/wl_window.c
+++ b/src/wl_window.c
@@ -2538,12 +2538,6 @@ void _glfwGetCursorPosWayland(_GLFWwindow* window, double* xpos, double* ypos)
         *ypos = window->wl.cursorPosY;
 }
 
-void _glfwSetCursorPosWayland(_GLFWwindow* window, double x, double y)
-{
-    _glfwInputError(GLFW_FEATURE_UNAVAILABLE,
-                    "Wayland: The platform does not support setting the cursor position");
-}
-
 void _glfwSetCursorModeWayland(_GLFWwindow* window, int mode)
 {
     _glfwSetCursorWayland(window, window->wl.currentCursor);
@@ -2861,6 +2855,26 @@ static void unconfinePointer(_GLFWwindow* window)
     window->wl.confinedPointer = NULL;
 }
 
+void _glfwSetCursorPosWayland(_GLFWwindow* window, double x, double y)
+{
+    if (window->wl.lockedPointer)
+    {
+        zwp_locked_pointer_v1_set_cursor_position_hint(window->wl.lockedPointer,
+                                                       wl_fixed_from_double(x),
+                                                       wl_fixed_from_double(y));
+    }
+    else
+    {
+        lockPointer(window);
+        zwp_locked_pointer_v1_set_cursor_position_hint(window->wl.lockedPointer,
+                                                       wl_fixed_from_double(x),
+                                                       wl_fixed_from_double(y));
+        unlockPointer(window);
+    }
+    window->wl.cursorPosX = x;
+    window->wl.cursorPosY = y;
+}
+
 void _glfwSetCursorWayland(_GLFWwindow* window, _GLFWcursor* cursor)
 {
     if (!_glfw.wl.pointer)
-- 
2.42.0

